### PR TITLE
[Feature](cloud) Add cloud report for clean up expired tablets

### PR DESCRIPTION
**Pull request by deardeng**
## Proposed changes

1. add tablet report to fe in cloud
2. clean up expired tablet in be

Issue Number: close #xxx

### CODE REVIEW

**Comment by deardeng** [comment]
wait regression case

**Comment by deardeng** [comment]
run buildall

File: be/src/agent/agent_server.cpp [hunk 1]
```diff
@@ -193,7 +194,7 @@ void AgentServer::start_workers(StorageEngine& engine, ExecEnv* exec_env) {
             "REPORT_DISK_STATE", _master_info, config::report_disk_state_interval_seconds, [&engine, &master_info = _master_info] { report_disk_callback(engine, master_info); }));
 
     _report_workers.push_back(std::make_unique<ReportWorker>(
-            "REPORT_OLAP_TABLE", _master_info, config::report_tablet_interval_seconds,[&engine, &master_info = _master_info] { report_tablet_callback(engine, master_info); }));
```
Comment (yujun777) [inline be/src/agent/agent_server.cpp:196]: ReportAction  use  report worker's name.  service/http_service.cpp need modify REPORT_OLAP_TABLE too.

File: be/src/agent/agent_server.cpp [hunk 2]
```diff
@@ -211,13 +212,25 @@ void AgentServer::cloud_start_workers(CloudStorageEngine& engine, ExecEnv* exec_
...
+
     _report_workers.push_back(std::make_unique<ReportWorker>(
             "REPORT_TASK", _master_info, config::report_task_interval_seconds,
             [&master_info = _master_info] { report_task_callback(master_info); }));
 
     _report_workers.push_back(std::make_unique<ReportWorker>(
             "REPORT_DISK_STATE", _master_info, config::report_disk_state_interval_seconds,
             [&engine, &master_info = _master_info] { report_disk_callback(engine, master_info); }));
+
+    if (config::enable_cloud_tablet_report) {
```
Comment (yujun777) [inline be/src/agent/agent_server.cpp:227]: if modify this param, need restart be to take effect ?

File: be/src/agent/agent_server.cpp [hunk 3]
```diff
@@ -193,7 +194,7 @@ void AgentServer::start_workers(StorageEngine& engine, ExecEnv* exec_env) {
             "REPORT_DISK_STATE", _master_info, config::report_disk_state_interval_seconds, [&engine, &master_info = _master_info] { report_disk_callback(engine, master_info); }));
 
     _report_workers.push_back(std::make_unique<ReportWorker>(
-            "REPORT_OLAP_TABLE", _master_info, config::report_tablet_interval_seconds,[&engine, &master_info = _master_info] { report_tablet_callback(engine, master_info); }));
+            "REPORT_OLAP_TABLET", _master_info, config::report_tablet_interval_seconds,[&engine, &master_info = _master_info] { report_tablet_callback(engine, master_info); }));
```
Comment (yujun777) [inline be/src/agent/agent_server.cpp:197]: ReportAction use report worker's name. service/http_service.cpp need modify REPORT_OLAP_TABLE too.

what's more, service/http_service.cpp contains a api  /api/report/tablet for not-cloud mode,  also add this api to cloud mode ?

File: be/src/agent/task_worker_pool.cpp [hunk 4]
```diff
@@ -1134,6 +1140,46 @@ void report_tablet_callback(StorageEngine& engine, const TMasterInfo& master_inf
...
+    for (int i = 0; i < 5; i++) {
+        request.tablets.clear();
+        report_version = s_report_version;
+        engine.tablet_mgr().build_all_report_tablets_info(&request.tablets, &tablet_num);
+        if (report_version == s_report_version) {
+            break;
+        }
+    }
+
+    if (report_version < s_report_version) {
```
Comment (yujun777) [inline be/src/agent/task_worker_pool.cpp:1167]: do cloud tablet report need check report_version ? maybe a stale report is ok.

Comment (deardeng) [reply be/src/agent/task_worker_pool.cpp:1167]: 目前没用到report version，留着看看以后会不会用？

File: be/src/agent/agent_server.cpp (hunk 2 above)
Comment (deardeng) [reply be/src/agent/agent_server.cpp:227]: fixed

**Comment by deardeng** [comment]
run buildall

**Review by yujun777 — APPROVED** [review]
LGTM

File: gensrc/thrift/HeartbeatService.thrift [hunk 5]
```diff
@@ -41,6 +41,8 @@ struct TMasterInfo {
     9: optional list<TFrontendInfo> frontend_infos
     10: optional string meta_service_endpoint;
     11: optional string cloud_unique_id;
+    // See configuration item Config.java rehash_tablet_after_be_dead_seconds for meaning
+    12: optional i64 cloud_tablet_report_exceed_time_limit;
```
Comment (gavinchou) [inline gensrc/thrift/HeartbeatService.thrift:45]: does it mean "timeout" or "interval"?

File: fe/fe-core/src/main/java/org/apache/doris/cloud/master/CloudReportHandler.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                .getSnapshotTabletsInPrimaryAndSecondaryByBeId(backendId);
+
+        Set<Long> tabletIdsInBe = backendTablets.keySet();
+        // handle (be - meta)
+        Set<Long> tabletIdsNeedDrop = diffTablets(tabletIdsInFe, tabletIdsInBe);
+        // drop agent task
+        deleteFromBackend(backendId, tabletIdsNeedDrop);
+
+        Backend be = Env.getCurrentSystemInfo().getBackend(backendId);
+        LOG.info("finished to handle task report from backend {}-{}, "
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/master/CloudReportHandler.java:56]: we can also print the first N, say 100, tablet ids in the log.

File: fe/fe-core/src/main/java/org/apache/doris/cloud/master/CloudReportHandler.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        Set<Long> result = new HashSet<>(tabletIdsInBe);
+        result.removeAll(tabletIdsInFe);
+        return result;
+    }
+
+    private static void deleteFromBackend(long backendId, Set<Long> tabletIdsWillDrop) {
+        int deleteFromBackendCounter = 0;
+        AgentBatchTask batchTask = new AgentBatchTask();
+        for (Long tabletId : tabletIdsWillDrop) {
+            if (LOG.isDebugEnabled()) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/master/CloudReportHandler.java:77]: redundant log

File: be/src/olap/txn_manager.cpp [hunk 8]
```diff
@@ -34,6 +34,7 @@
 #include "common/config.h"
 #include "common/logging.h"
 #include "common/status.h"
+#include "olap/base_tablet.h"
```
Comment (gavinchou) [inline be/src/olap/txn_manager.cpp:37]: ?

File: be/src/olap/tablet_manager.cpp [hunk 9]
```diff
@@ -41,6 +41,7 @@
 #include "gutil/strings/strcat.h"
 #include "gutil/strings/substitute.h"
 #include "io/fs/local_file_system.h"
+#include "olap/base_tablet.h"
```
Comment (gavinchou) [inline be/src/olap/tablet_manager.cpp:44]: ?

File: be/src/cloud/cloud_tablet_mgr.h [hunk 10]
```diff
@@ -65,6 +68,11 @@ class CloudTabletMgr {
                                        std::vector<std::shared_ptr<CloudTablet>>* tablets,
                                        int64_t* max_score);
 
+    void build_all_report_tablets_info(std::map<TTabletId, TTablet>* tablets_info,
```
Comment (gavinchou) [inline be/src/cloud/cloud_tablet_mgr.h:71]: add comment

File: be/src/agent/task_worker_pool.cpp [hunk 11]
```diff
@@ -1134,6 +1140,46 @@ void report_tablet_callback(StorageEngine& engine, const TMasterInfo& master_inf
...
+    }
+
+    TReportRequest request;
+    request.__set_backend(BackendOptions::get_local_backend());
+    request.__isset.tablets = true;
+
+    increase_report_version();
+    uint64_t report_version;
+    uint64_t tablet_num = 0;
+    for (int i = 0; i < 5; i++) {
```
Comment (gavinchou) [inline be/src/agent/task_worker_pool.cpp:1158]: why for 5 times?

File: be/src/agent/task_worker_pool.cpp [hunk 12]
```diff
@@ -1134,6 +1140,46 @@ void report_tablet_callback(StorageEngine& engine, const TMasterInfo& master_inf
...
+        random_sleep(5);
+    }
+
+    TReportRequest request;
+    request.__set_backend(BackendOptions::get_local_backend());
+    request.__isset.tablets = true;
+
+    increase_report_version();
+    uint64_t report_version;
+    uint64_t tablet_num = 0;
```
Comment (gavinchou) [inline be/src/agent/task_worker_pool.cpp:1157]: total_num_tablets

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by deardeng** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by gavinchou — APPROVED** [review]

**Pull request merged**